include/openthread/platform/radio.h

```diff
@@ -822,11 +822,16 @@ otError otPlatRadioReceive(otInstance *aInstance, uint8_t aChannel);
  *                        still actively receiving a frame. In the latter case
  *                        the radio SHALL be kept in reception mode until frame
  *                        reception has either succeeded or failed.
+ * @param[in]  aSlotId    The receive window slot ID.
  *
  * @retval OT_ERROR_NONE    Successfully scheduled receive window.
  * @retval OT_ERROR_FAILED  The receive window could not be scheduled.
  */
-otError otPlatRadioReceiveAt(otInstance *aInstance, uint8_t aChannel, uint32_t aStart, uint32_t aDuration);
+otError otPlatRadioReceiveAt(otInstance *aInstance,
+                             uint8_t     aChannel,
+                             uint32_t    aStart,
+                             uint32_t    aDuration,
+                             uint8_t     aSlotId);
```

Comment:
I haven't had the time to review the rest of this PR, but I noticed this change in `radio.h`. 

This change modifies an existing `otPlatRadioReceiveAt()` API (by adding a new parameter). Our public and platform APIs are our contract with users and platforms, and we generally keep them unchanged to ensure backward compatibility. We can always add new APIs, but if we want to change an existing one, we need a strong justification (after exploring all other options).

Some questions/thoughts:

- Perhaps this is answered in the code or spec, but I'm curious why we need the `aSlotId`. What does it do? How should it be used by the radio platform? (It might be helpful to document this in the API regardless.)
- Can we avoid adding `aSlotId` and use the existing API as is instead? For example, if it's used to determine or shift the receive window, can it be calculated within the OT sub-MAC?

If it's necessary to inform the radio about the Slot ID, could we convey this information in a different way?

- Add a new API to set the slot ID to use?
- Or add a new API like `otPlatRadioReceiveAtSlotId()`?

Reply:
Hi @abtink, the need of receive slots IDs comes from this (optional) part of the spec:

> The WED sleep-and-monitor state can also be combined by a Thread Device with other forms of
> connectivity, such as being connected as a **SSED** or SED to a regular Thread Router, if such
> Router is available.

So basically we need to be able to handle both CSL and WED slots at the same time (both scheduled independently).

I was implemented by extending `otPlatRadioReceiveAt()` due to code efficiency, but I believe something like `otPlatRadioReceiveAtNextSlotId()` could work in order to avoid breaking existing API.

Reply:
Thanks @edmont.

The existing `otPlatRadioReceiveAt()` behavior is clear: we're instructing the radio to receive starting at a specific time, for a given duration, and on a given channel.

With the updated definition and the new `aSlotId` input, what is the radio platform expected to do?

* Does it mean it should receive at `aStart` for `aDuration` **AND** also at the given `aSlotId`?
* If these are two separate, independent operations, defining a separate API would be much cleaner than combining them into one.

-----

Additionally, do we need the radio platform to be aware of the "slot ID"? Could we have the `SubMac` code track this and use the existing `otPlatRadioReceiveAt()`, passing it the appropriate `aStart` and `aDuration` corresponding to the slot ID? The fewer dependencies and requirements we place on the radio platform, the more testable and portable the code will be.

Reply:
When both WED listening and CSL reception are happening concurrently:
- Both kind of receive slots can be scheduled independently by `SubMac` (i.e., we need the radio to be capable of handling two delayed receptions at the same time).
- While both are already scheduled, some condition can occur (for instance, CSL period change) which would trigger a re-scheduling of non-expired slot.
- In this case we need to be able to cancel such slot but not the other one: thus the need of IDs.
- In our current radio driver code for the prototype we are always cancelling given slot by ID whenever `otPlatRadioReceiveAt()` schedules it.
---
Since we are never overlapping slots (see `GetNeededShift()`), in theory it would be possible to support both functionalities without new radio capabilities (always a single slot scheduled at a given time). But this approach would have two major issues:
- Being able to rapidly re-schedule a slot (would be problematic for RCP architecture).
- Mixing or intelligently handling WED and CSL state machines.
---
For now it seems a way forward could be:
1. Remove support for SSED + WED from this PR (keep it simple for now). `otLinkWedListenSetEnabled` would return an error if CSL Period is not 0 (and vice-versa).
2. Separately add `OT_RADIO_CAPS_TWO_RECEIVE_SLOTS` capability handling.
3. Add support for SSED + WED for devices supporting such capability (parts of this PR).
4. For the future: add support for SSED + WED for platforms without the capability.

Thoughts?

Reply:
Thanks, @edmont, for your response. Keeping it simple for now sounds like a good plan.

> But this approach would have two major issues:
>
> - Being able to rapidly reschedule a slot (would be problematic for RCP architecture).
> - Mixing or intelligently handling WED and CSL state machines.

I agree with these, but if I understand correctly, we'll need to handle these situations regardless. The question is whether to delegate this to the radio driver or handle it within the `SubMac` (or `Mac`) implementation. I generally prefer to keep more functionality in the common part of the code to promote reusability and reduce the number of combination cases that need testing.

Reply:
@abtink I think one argument to delegate this to the radio driver is to achieve a shorter turnaround time between scheduling a CSL slot and a wake-up slot (and vice versa). Depending on the architecture, it may take a few milliseconds to schedule another slot in SubMac, by which the wake-up slot would have to be shifted in case the wake-up slot occurs right after the CSL slot, reducing the probability of receiving a wake-up frame.

Reply:
Good point. To improve this, I would suggest adding new radio platform APIs to allow scheduling multiple receive time windows:

-   A simple approach would be to use the existing `otPlatRadioReceiveAt()` for CSL only and add a new radio API specifically for scheduling wake-ups.
-   A potentially better approach would be to add a more generic way to schedule multiple receive windows and provide a way to update or cancel each individually. Perhaps something like this:

```c++
typedef struct otPlatRadioRxSchedule
{
    uint8_t  mChannel;   ///< Channel to receive on.
    uint32_t mStartTime; ///< Start time of the receive window, based on the local radio clock (`otPlatRadioGetNow()`).
    uint32_t mDuration;  ///< The receive window duration in microseconds.
} otPlatRadioRxSchedule;

/**
 * Schedule a radio reception window at a specific time and duration.
 *
 * @param[in]  aInstance    The OpenThread instance structure.
 * @param[in]  aIndex       The schedule index (can be used to update or cancel a previously scheduled time).
 * @param[in]  aRxSchedule  The schedule (channel, start time, and duration).
 *
 * @retval OT_ERROR_NONE     Successfully added the schedule.
 * @retval OT_ERROR_FAILED   Failed to add the schedule.
 */
otError otPlatRadioSetRxSchedule(otInstance *aInstance, uint8_t aIndex, const otPlatRadioRxSchedule *aRxSchedule);

/**
 * Cancel a previously scheduled radio reception.
 *
 * @param[in]  aInstance    The OpenThread instance structure.
 * @param[in]  aIndex       The schedule index.
 */
void otPlatRadioCancelRxSchedule(otInstance *aInstance, uint8_t aIndex);
```

What do you think?

Reply:
By the way, we could defer any such discussion/changes to future PRs (as @edmont suggested, we can keep this PR simple and disallow the CSL and WED combination for now).